podvm/qcow2/build-se-image.sh

```diff
@@ -0,0 +1,165 @@
+if [ "${SE_BOOT:-0}" != "1" ]; then
+    exit 0
+elif [ "${ARCH}" != "s390x" ]; then
+    echo "Build SE podvm image for $ARCH is not support yet"
+    exit 0
+fi
+echo "Building SE podvm image for $ARCH" 
+echo "Finding host key files" 
+host_keys=""
+for i in $(ls /tmp/files/*.crt); do
+    echo "found host key file: \"${i}\""
+    host_keys+="-k ${i} "
+done
+[[ -z $host_keys ]] && echo "Didn't find host key files, please download host key files to `files` folder " && exit 1
+export host_keys
+export workdir=$(pwd)
+export disksize=100G
+export device=$(lsblk |grep 100G | awk '{ print $1 }')
+# /dev/vda or /dev/vdb
+export tmp_nbd="/dev/$device"
+export dst_mnt=$workdir/dst_mnt
+# /dev/vda1 or /dev/vdb1
+export src_nbd=$(mount |grep "on / type ext4 (rw,relatime)" | awk '{ print $1 }')
+
+echo "Creating boot-se and root partitions"
+
+sudo parted -a optimal ${tmp_nbd} mklabel gpt \
+    mkpart boot-se ext4 1MiB 256MiB \
+    mkpart root 256MiB ${disksize} \
+    set 1 boot on
+
+echo "Waiting for the two partitions to show up"
+while true; do
+sleep 1
+[ -e ${tmp_nbd}2 ] && break
+done
+
+
+echo "Formatting boot-se partition"
+sudo mke2fs -t ext4 -L boot-se ${tmp_nbd}1
+export boot_uuid=$(sudo blkid ${tmp_nbd}1 -s PARTUUID -o value)
+
+echo "Setting up encrypted root partition"
+sudo mkdir ${workdir}/rootkeys
+sudo mount -t tmpfs rootkeys ${workdir}/rootkeys
+sudo dd if=/dev/random of=${workdir}/rootkeys/rootkey.bin bs=1 count=64 &> /dev/null
+echo YES | sudo cryptsetup luksFormat --type luks2 ${tmp_nbd}2 --key-file ${workdir}/rootkeys/rootkey.bin
+echo "Setting luks name for root partition"
+export LUKS_NAME="luks-$(sudo blkid -s UUID -o value ${tmp_nbd}2)"
+echo "luks name is: $LUKS_NAME"
+sudo cryptsetup open ${tmp_nbd}2 $LUKS_NAME --key-file ${workdir}/rootkeys/rootkey.bin
+
+# TOOTOTO
+echo "Copying root partition with dd"
+sudo dd if="${src_nbd}" of="/dev/mapper/${LUKS_NAME}" bs=$((1024*1024))
```

Comment:
I tied use the `tar` command to copy the running root filesystem to another mount device, but I meet follow errors:
```
peerpod@ubuntu:~$ sudo tar -cf - "${tar_opts[@]}" --sort=none -C "$src_mnt" . | tar -xf - "${tar_opts[@]}" --preserve-order  -C "$dst_mnt"
tar: ./lib: Cannot create symlink to ‘usr/lib’: Permission denied
tar: ./boot: Cannot mkdir: Permission denied
tar: ./boot: Cannot mkdir: Permission denied
tar: ./boot/System.map-5.4.0-147-generic: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: Permission denied
...
```

Reply:
Copying a system filesystem with `tar` needs a lot of options. I think we need at least [`--one-file-system`](https://serverfault.com/a/561735) in addition to the options I used below.

https://github.com/confidential-containers/cloud-api-adaptor/blob/e7183af5889436de4b750d6fb0c79fe44c00ec97/ibmcloud/image/build.sh#L287-L289

I think `--one-file-system` is necessary, since various additional filesystems such as `/tmp` and `/proc` are mounted in a running Linux operating system.

You also need use `sudo` to run the second `tar`.

Reply:
I think the `tar` approach is safer than the `fsfreeze` approach, if it is functional.

Reply:
Thanks @yoheiueda ,
I updated the codes to use `tar` approach with `sudo mount --bind -o ro / ${src_mnt}`, the tar command output:
```
tar: ./var/snap/lxd/common/lxd/unix.socket: socket ignored
```
The built out se-image seems bigger than the `dd` approach, it's 3.5G
```
-rw-r--r-- 1 root root 3.5G May  8 06:27 se-podvm-2413ed0-dirty-s390x.qcow2
```

The built out qcow2 image works as expected, e2e test are passed with it.

Reply:
@liudalibj Thank you for the update!

I thought the se-image would be 100GB, since luksFormat leads to allocate all disk space of the QCOW2 image. I didn't think QCOW2's copy-on-write mechanism works for encrypted disks. Do you have any idea how it works?

Supposing lazy allocation works for encrypted disks, the difference between the two approaches may be the actual partition sizes. I think the source partition of the `dd` command was not 100GB. Right?

https://github.com/confidential-containers/cloud-api-adaptor/blob/bc64906141e15de4388ef53a281a96ed02b99f44/podvm/qcow2/ubuntu/variables.pkr.hcl#L16-L19

I think 100GB is a limitation for IBM Cloud VPC, and we don't need full 100GB space for peer pod VM components. 
A possible work around is to create a 6GB partition, and encrypted it with LUKS. We can keep the remaining spare unused.

Reply:
@yoheiueda I don't know how it works that why with luksFormat we didn't get one 100G qcow2 file.

And I tried your suggestion, to make the root partition to 6G and then encrypted it.
The final output qcow2 file is reduced to 1.7 G, thanks!
```
-rw-r--r-- 1 root root 1.7G May 10 06:21 se-podvm-0b154ef-dirty-s390x.qcow2
```
So we have one unused `data` partition after create one SE VSI with the built out image.
```
peerpod@se-test-38:~$ lsblk
NAME                                          MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINT
loop0                                           7:0    0 54.6M  1 loop  /snap/core20/1881
loop1                                           7:1    0 77.8M  1 loop  /snap/lxd/24063
loop2                                           7:2    0 49.6M  1 loop  /snap/snapd/19126
vda                                           252:0    0  100G  0 disk
├─vda1                                        252:1    0  255M  0 part  /boot-se
├─vda2                                        252:2    0    6G  0 part
│ └─luks-aff936d5-0d05-4ea3-b038-192ba4df6f41 253:0    0    6G  0 crypt /
└─vda3                                        252:3    0 86.9G  0 part
vdb                                           252:16   0  366K  0 disk
vdc                                           252:32   0   44K  0 disk
peerpod@se-test-38:~$ exit
```